File: R5ProObjectiveCExamples/R5ProObjectiveCExamples/Tests/BaseTest.m
//
//  BaseTest.m
//  R5ProObjectiveCExamples
//
//  Created by David Heimann on 6/6/17.
//  Copyright © 2017 Infrared5. All rights reserved.
//

#import <UIKit/UIKit.h>
#import "BaseTest.h"
#import "ALToastView.h"
#import "Testbed.h"

@interface BaseTest ()

@end

@implementation BaseTest

-(id)init{
    return [super initWithNibName:nil bundle:nil];
    _shouldAutorotate = YES;
    _supportedInterfaceOrientations = UIInterfaceOrientationMaskAll;
}

-(id)initWithCoder:(NSCoder *)aDecoder{
    return [super initWithCoder:aDecoder];
    _shouldAutorotate = YES;
    _supportedInterfaceOrientations = UIInterfaceOrientationMaskAll;
}

-(void) onR5StreamStatus:(R5Stream *)stream withStatus:(int)statusCode withMessage:(NSString *)msg{
    NSLog(@"Status: %s", r5_string_for_status(statusCode));
    NSString* s = [NSString stringWithFormat:@"Status: %s (%@)", r5_string_for_status(statusCode), msg];
    
    [ALToastView toastInView:self.view withText:s];
}

-(void) closeTest{
    
    NSLog(@"closing view");
    
    if( _publishStream != nil ){
        [_publishStream stop];
    }
    
    if( _subscribeStream != nil ){
        [_subscribeStream stop];
    }
    
    [self removeFromParentViewController];
}

-(R5Configuration*) getConfig{
    R5Configuration* config = [[R5Configuration alloc] init];
    config.host = [Testbed getParameter:@"host"];
    config.port = (int)[[Testbed getParameter:@"port"] integerValue];
    config.contextName = [Testbed getParameter:@"context"];
    config.protocol = 1;
    config.buffer_time = [[Testbed getParameter:@"buffer_time"] floatValue];
    config.licenseKey = [Testbed getParameter:@"license_key"];
    
    return config;
}

-(void) setupPublisher:(R5Connection*)connection{
    _publishStream = [[R5Stream alloc] initWithConnection:connection];
    _publishStream.delegate = self;
    
    if( [[Testbed getParameter:@"video_on"] boolValue] ){
        // Attach the video from camera to stream
        AVCaptureDevice *videoDevice = [[AVCaptureDevice devicesWithMediaType:AVMediaTypeVideo] lastObject];
        
        R5Camera *camera = [ [R5Camera alloc] initWithDevice:videoDevice andBitRate:[[Testbed getParameter:@"bitrate"] intValue] ];
        
        camera.width = [[Testbed getParameter:@"camera_width"] intValue];
        camera.height = [[Testbed getParameter:@"camera_height"] intValue];
        camera.orientation = 90;
        [self.publishStream attachVideo:camera];
    }
    if ( [[Testbed getParameter:@"audio_on"] boolValue] ) {
        AVCaptureDevice* audioDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeAudio];
        R5Microphone* microphone = [[R5Microphone alloc] initWithDevice:audioDevice];
        microphone.bitrate = 32;
        microphone.device = audioDevice;
        NSLog(@"Got device %@", audioDevice);
        [_publishStream attachAudio:microphone];
    }
}

-(R5VideoViewController*) setupDefaultR5VideoViewController{
    
    R5VideoViewController* r5View = [self getNewR5VideoViewController:[self view].frame];
    [self addChildViewController:r5View];
    
    [[self view] addSubview:r5View.view];
    [r5View setFrame:[self view].bounds];
    
    [r5View showPreview:YES];
    [r5View showDebugInfo:[[Testbed getParameter:@"debug_view"] boolValue]];
    
    _currentView = r5View;
    
    return _currentView;
}

-(R5VideoViewController*) getNewR5VideoViewController:(CGRect)rect{
    
    UIView* view = [[UIView alloc] initWithFrame:rect];
    
    R5VideoViewController* r5View = [[R5VideoViewController alloc] init];
    r5View.view = view;
    
    return r5View;
}

-(void) viewDidLayoutSubviews{
    [super viewDidLayoutSubviews];
    
    if(_currentView != nil){
        [_currentView setFrame: [self view].frame];
    }
}

-(void) viewDidLoad{
    [super viewDidLoad];
    
    [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted){
        
    }];
    
    r5_set_log_level(r5_log_level_debug);
    
    [self view].autoresizesSubviews = NO;
}

-(void) viewDidAppear:(BOOL)animated{
    
    //this is just to have a white background to the example
    UIView* backView = [[UIView alloc] initWithFrame:[self view].frame];
    backView.backgroundColor = [UIColor whiteColor];
    [[self view] addSubview:backView];
}

@end

File: R5ProTestbed/Tests/BaseTest.swift
//
//  BaseTest.swift
//  R5ProTestbed
//
//  Created by Andy Zupko on 12/16/15.
//  Copyright © 2015 Infrared5, Inc. All rights reserved.
// 
//  The accompanying code comprising examples for use solely in conjunction with Red5 Pro (the "Example Code") 
//  is  licensed  to  you  by  Infrared5  Inc.  in  consideration  of  your  agreement  to  the  following  
//  license terms  and  conditions.  Access,  use,  modification,  or  redistribution  of  the  accompanying  
//  code  constitutes your acceptance of the following license terms and conditions.
//  
//  Permission is hereby granted, free of charge, to you to use the Example Code and associated documentation 
//  files (collectively, the "Software") without restriction, including without limitation the rights to use, 
//  copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit 
//  persons to whom the Software is furnished to do so, subject to the following conditions:
//  
//  The Software shall be used solely in conjunction with Red5 Pro. Red5 Pro is licensed under a separate end 
//  user  license  agreement  (the  "EULA"),  which  must  be  executed  with  Infrared5,  Inc.   
//  An  example  of  the EULA can be found on our website at: https://account.red5pro.com/assets/LICENSE.txt.
// 
//  The above copyright notice and this license shall be included in all copies or portions of the Software.
//  
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,  INCLUDING  BUT  
//  NOT  LIMITED  TO  THE  WARRANTIES  OF  MERCHANTABILITY, FITNESS  FOR  A  PARTICULAR  PURPOSE  AND  
//  NONINFRINGEMENT.   IN  NO  EVENT  SHALL INFRARED5, INC. BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, 
//  WHETHER IN  AN  ACTION  OF  CONTRACT,  TORT  OR  OTHERWISE,  ARISING  FROM,  OUT  OF  OR  IN CONNECTION 
//  WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

//

import UIKit
import R5Streaming

enum AccessError: Error {
    case error(message: String)
}

extension AccessError: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .error:
            return NSLocalizedString("Unable to fetch stream for action subscribe.", comment: "Access Error")
        }
    }
}

@objc(BaseTest)
class BaseTest: UIViewController , R5StreamDelegate {
    
    func onR5StreamStatus(_ stream: R5Stream!, withStatus statusCode: Int32, withMessage msg: String!) {
        let s =  String(format: "Status: %s (%@)",  r5_string_for_status(statusCode), msg)
        NSLog(s)
        ALToastView.toast(in: self.view, withText:s)
        
        if (Int(statusCode) == Int(r5_status_disconnected.rawValue)) {
            self.cleanup()
        }
        else if (Int(statusCode) == Int(r5_status_video_render_start.rawValue)) {
            NSLog("SUPPORT-482 %@", msg);
        }
    }
    
    var shouldClose : Bool = true
    var currentView : R5VideoViewController? = nil
    var publishStream : R5Stream? = nil
    var subscribeStream : R5Stream? = nil
    
    required init () {
       
        super.init(nibName: nil, bundle: nil)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    func cleanup () {
        if( self.publishStream != nil ) {
            self.publishStream!.client = nil
            self.publishStream?.delegate = nil
            self.publishStream = nil
        }
        
        if( self.subscribeStream != nil ) {
            self.subscribeStream!.client = nil
            self.subscribeStream?.delegate = nil
            self.subscribeStream = nil
        }
        self.removeFromParent()
    }
    
    func closeTest(){
        
        NSLog("closing view")

        if( self.publishStream != nil ){
            self.publishStream!.stop()
        }
        
        if( self.subscribeStream != nil ){
            self.subscribeStream!.stop()
        }
        
        // Moved to status disconnect, due to publisher emptying queue buffer on bad connections.
//        self.removeFromParentViewController()
    }
    
    func getConfig()->R5Configuration{
        // Set up the configuration
        let config = R5Configuration()
        config.host = Testbed.getParameter(param: "host") as! String
        config.port = Int32(Testbed.getParameter(param: "port") as! Int)
        config.contextName = Testbed.getParameter(param: "context") as! String
        config.`protocol` = Int32(r5_rtsp.rawValue);
        config.buffer_time = (Testbed.getParameter(param: "buffer_time")?.floatValue)!
        config.licenseKey = Testbed.getParameter(param: "license_key") as! String
        return config
    }
    
    
    override func viewDidLayoutSubviews() {
        
        super.viewDidLayoutSubviews()
        
        if(currentView != nil){
            
            currentView?.setFrame(view.frame);
        }
        
    }
    
    func setupPublisher(connection: R5Connection){
        
        self.publishStream = R5Stream(connection: connection)
        self.publishStream!.delegate = self
        
        if(Testbed.getParameter(param: "video_on") as! Bool){
            // Attach the video from camera to stream
            let videoDevice = AVCaptureDevice.devices(for: AVMediaType.video).last as? AVCaptureDevice
            
            let camera = R5Camera(device: videoDevice, andBitRate: Int32(Testbed.getParameter(param: "bitrate") as! Int))
           
            camera?.width = Int32(Testbed.getParameter(param: "camera_width") as! Int)
            camera?.height = Int32(Testbed.getParameter(param: "camera_height") as! Int)
            camera?.fps = Int32(Testbed.getParameter(param: "fps") as! Int)
            camera?.orientation = 90
            self.publishStream!.attachVideo(camera)
        }
        if(Testbed.getParameter(param: "audio_on") as! Bool){
            // Attach the audio from microphone to stream
            let audioDevice = AVCaptureDevice.default(for: AVMediaType.audio)
            let microphone = R5Microphone(device: audioDevice)
            microphone?.bitrate = 32
            NSLog("Got device %@", String(describing: audioDevice?.localizedName))
            self.publishStream!.attachAudio(microphone)
        }

    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        AVAudioSession.sharedInstance().requestRecordPermission { (gotPerm: Bool) -> Void in
            
        };
        
        r5_set_log_level((Int32)(r5_log_level_debug.rawValue))
        
        self.view.autoresizesSubviews = false
        
    }
    
    override func viewDidAppear(_ animated: Bool) {
        
        super.viewDidAppear(animated)
        
        //this is just to have a white background to the example
        let backView : UIView = UIView(frame: self.view.frame);
        backView.backgroundColor = UIColor.white;
        self.view.addSubview(backView);
        
    }
    
    func getPublishRecordType () -> R5RecordType {
        var type = R5RecordTypeLive
        if Testbed.getParameter(param: "record_on") as! Bool {
            type = R5RecordTypeRecord
            if Testbed.getParameter(param: "append_on") as! Bool {
                type = R5RecordTypeAppend
            }
        }
        return type
    }
    
    func setupDefaultR5VideoViewController() -> R5VideoViewController{
        
        currentView = getNewR5VideoViewController(rect: self.view.frame)
        
        self.addChild(currentView!)
        self.view.addSubview(currentView!.view)
        
        currentView?.setFrame(self.view.bounds)
        
        currentView?.showPreview(true)

        currentView?.showDebugInfo(Testbed.getParameter(param: "debug_view") as! Bool)
        
        return currentView!
    }
    
    func getNewR5VideoViewController(rect : CGRect) -> R5VideoViewController {
        
        let view : UIView = UIView(frame: rect)
        
        var r5View : R5VideoViewController
        r5View = R5VideoViewController.init()
        r5View.view = view;
        
        return r5View;
            
    }
    
    open override var shouldAutorotate:Bool {
        get {
            return true
        }
    }
    
    open override var supportedInterfaceOrientations: UIInterfaceOrientationMask {
        get {
            return [UIInterfaceOrientationMask.all]
        }
    }

}
